javaslang/src/main/java/javaslang/collection/BitSet.java

```diff
@@ -0,0 +1,1024 @@
+            } else {
+                return false;
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return Collections.hash(this);
+        }
+    }
+
+    class BitSet1<T> extends AbstractBitSet<T> {
+
+        private static final long serialVersionUID = 1L;
+
+        private final long elements;
+
+        BitSet1(Function1<Integer, T> fromInt, Function1<T, Integer> toInt, long elements) {
+            super(fromInt, toInt);
+            this.elements = elements;
+        }
+
+        @Override
+        int getWordsNum() {
+            return 1;
+        }
+
+        @Override
+        long[] copyExpand(int wordsNum) {
+            if(wordsNum < 1) {
+                wordsNum = 1;
+            }
+            long[] arr = new long[wordsNum];
+            arr[0] = elements;
+            return arr;
+        }
+
+        @Override
+        long getWord(int index) {
+            if (index == 0) {
+                return elements;
+            } else {
+                return 0L;
+            }
+        }
+
+        @Override
+        public T head() {
+            if (elements != 0) {
+                return fromInt.apply(Long.numberOfTrailingZeros(elements));
+            }
+            throw new NoSuchElementException("head of empty BitSet");
+        }
+    }
+
+    class BitSet2<T> extends AbstractBitSet<T> {
+
+        private static final long serialVersionUID = 1L;
+
+        private final long elements1, elements2;
+
+        BitSet2(Function1<Integer, T> fromInt, Function1<T, Integer> toInt, long elements1, long elements2) {
+            super(fromInt, toInt);
+            this.elements1 = elements1;
+            this.elements2 = elements2;
+        }
+
+        @Override
+        int getWordsNum() {
+            return 2;
+        }
+
+        @Override
+        long[] copyExpand(int wordsNum) {
+            if(wordsNum < 2) {
+                wordsNum = 2;
+            }
+            long[] arr = new long[wordsNum];
+            arr[0] = elements1;
+            arr[1] = elements2;
+            return arr;
+        }
+
+        @Override
+        long getWord(int index) {
+            if (index == 0) {
+                return elements1;
+            } else {
+                if (index == 1) {
+                    return elements2;
+                } else {
+                    return 0L;
+                }
+            }
+        }
+
+        @Override
+        public T head() {
+            if (elements1 != 0) {
+                return fromInt.apply(Long.numberOfTrailingZeros(elements1));
+            } else if (elements2 != 0) {
+                return fromInt.apply(BITS_PER_WORD + Long.numberOfTrailingZeros(elements2));
+            }
+            throw new NoSuchElementException("head of empty BitSet");
+        }
+    }
+
+    class BitSetN<T> extends AbstractBitSet<T> {
+
+        private static final long serialVersionUID = 1L;
+
+        private final long[] elements;
```

Comment:
... and the `shrink` method, but that still might not be enough reason. You know the details, I just have some ideas :).
The advantage of `Array` would be that you could `head`, `tail`, `drop`, `dropRight`, `take` etc. in constant time, instead of linear, as it is currently.